### PR TITLE
[Consensus] New logic for round duration calculation

### PR DESCRIPTION
**Pull request by durkmurder**
### Context

This PR implements new logic for calculating round duration as described [here](https://www.notion.so/dapperlabs/Architecture-for-Active-PaceMaker-2b400d5218d9440c8ec98323fb818d92#98ddf2d9b5314a7c909a6593b8815551). 

Main changes:
- Updated config to accept, `tmin`, `tmax`, and `k`. 
- `TimeoutController` keeps track of number of failed rounds
- `TimeoutController` calculates max exponent in constructor
- Removed initial timeout from config
- Updated tests

### CODE REVIEW

**Review by AlexHentschel — APPROVED** [review]
Great work. Mostly stylistic comments. 
The only aspects, which I would suggest revising are:
1. Thank you for visualizing the different functions we considered ([notion](https://www.notion.so/dapperlabs/Architecture-for-Active-PaceMaker-2b400d5218d9440c8ec98323fb818d92#fa25f07c01194f89a6b77c2152719b23)). This made me realize that I actually prefer 
    Screen Shot 2022-09-19 at 9 44 05 PM
   Because after reaching `hotstuffHappyPathRounds`, the timeout slowly increases instead of jumping. I think this creates a generally more intuitive and stable behaviour for a wide range of parameter settings. Assume we set `TimeoutIncrease = 2` and `HappyPathRounds = 6` This is quite similar to our default parameter setting. In the scenario where we have several successive round failures, the timeout would stay at its initial value for 6 rounds and then jump to 64x the initial value. And exactly that jump after `HappyPathRounds`  is not desired / expected. 
   See [this comment](https://github.com/onflow/flow-go/pull/3201#discussion_r974875851) for implementation-specific suggestions. 
2. I don't think `r` (or `roundsWithoutProgress` as you call it) should be allowed to increase much beyond the `maxExponent` (see  [comment](https://github.com/onflow/flow-go/pull/3201#discussion_r974885044))

File: consensus/hotstuff/pacemaker/timeout/controller.go [hunk 1]
```diff
@@ -7,15 +7,29 @@ import (
...
-// - on progress: decrease timeout by subtrahend `timeoutDecrease`
+// Controller implements a truncated exponential backoff which can be described with next formula:
+// duration = t_min * min(b ^ (r * θ(r-k)), t_max)
+// for practical purpose we will transform this formula into:
+// duration(r) = t_min * b ^ (min(r * θ(r-k)), c), where c = b(t_max / t_min).
+// In described formula:
+//   k - is number of rounds we expect during hot path, after failing this many rounds,
+//   we will start increasing timeouts.
+//   b - timeout increase factor
+//   r - number of failed rounds
```
[the remark this answers is not in the document]
Comment (jordanschalm) [reply consensus/hotstuff/pacemaker/timeout/controller.go:18]: We should align on this point and update the [notion doc](https://www.notion.so/dapperlabs/Architecture-for-Active-PaceMaker-2b400d5218d9440c8ec98323fb818d92#a86eb80c9c5d406283e1d85e61449a82) as well, currently it says:
> $r$  is distance between current round and last round where we have observed progress

File: consensus/hotstuff/pacemaker/timeout/controller.go [hunk 2]
```diff
@@ -7,15 +7,29 @@ import (
...
 )
 
-// Controller implements a timout with:
-// - on timeout: increase timeout by multiplicative factor `timeoutIncrease` (user-specified)
-//   this results in exponential growing timeout duration on multiple subsequent timeouts
-// - on progress: decrease timeout by subtrahend `timeoutDecrease`
+// Controller implements a truncated exponential backoff which can be described with next formula:
+// duration = t_min * min(b ^ (r * θ(r-k)), t_max)
+// for practical purpose we will transform this formula into:
+// duration(r) = t_min * b ^ (min(r * θ(r-k)), c), where c = b(t_max / t_min).
```
[the remark this answers is not in the document]
Comment (jordanschalm) [reply consensus/hotstuff/pacemaker/timeout/controller.go:13]: Note: this is missing the `(r-k)` component in the exponent, as suggested in the [overall review comment](https://github.com/onflow/flow-go/pull/3201#pullrequestreview-1112920009):
image

File: consensus/hotstuff/pacemaker/timeout/controller.go [hunk 3]
```diff
@@ -66,17 +83,27 @@ func (t *Controller) StartTimeout(view uint64) *model.TimerInfo {
...
+
+	exponent := math.Min(float64(t.roundsWithoutProgress*step), t.maxExponent)
+	duration := t.cfg.MinReplicaTimeout * math.Pow(t.cfg.TimeoutIncrease, exponent)
+	return time.Duration(duration * float64(time.Millisecond))
 }
 
 // OnTimeout indicates to the Controller that a view change was triggered by a TC (unhappy path).
 func (t *Controller) OnTimeout() {
-	t.cfg.ReplicaTimeout = math.Min(t.cfg.ReplicaTimeout*t.cfg.TimeoutIncrease, t.cfg.MaxReplicaTimeout)
+	t.roundsWithoutProgress++
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply consensus/hotstuff/pacemaker/timeout/controller.go:99]: Good one, totally missed that, very good suggestion

File: consensus/hotstuff/pacemaker/timeout/controller_test.go [hunk 4]
```diff
@@ -51,28 +49,37 @@ func Test_TimeoutInitialization(t *testing.T) {
...
+		if uint64(i) > happyPathRounds {
+			step = 1
+		}
 		assert.Equal(t,
 			tc.ReplicaTimeout().Milliseconds(),
-			int64(repTimeout*math.Pow(multiplicativeDecrease, float64(i))),
+			int64(minRepTimeout*math.Pow(multiplicativeIncrease, float64(i*step))),
 		)
 	}
 }
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply consensus/hotstuff/pacemaker/timeout/controller_test.go:85]: my bad, sorry

File: consensus/hotstuff/pacemaker/timeout/controller.go [hunk 5]
```diff
@@ -66,17 +83,27 @@ func (t *Controller) StartTimeout(view uint64) *model.TimerInfo {
...
-	return time.Duration(t.cfg.ReplicaTimeout * 1e6)
+	// piecewise function
+	step := uint64(0)
+	if t.roundsWithoutProgress > t.cfg.HappyPathRounds {
+		step = 1
+	}
+
+	exponent := math.Min(float64(t.roundsWithoutProgress*step), t.maxExponent)
+	duration := t.cfg.MinReplicaTimeout * math.Pow(t.cfg.TimeoutIncrease, exponent)
+	return time.Duration(duration * float64(time.Millisecond))
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply consensus/hotstuff/pacemaker/timeout/controller.go:94]: good observation

File: consensus/hotstuff/pacemaker/timeout/controller_test.go [hunk 6]
```diff
@@ -95,19 +102,18 @@ func Test_MinCutoff(t *testing.T) {
...
+		happyPathRounds,
 		0)
 	if err != nil {
 		t.Fail()
 	}
 	tc := NewController(c)
 
-	for i := 1; i <= 50; i += 1 {
-		tc.OnTimeout() // after already 7 iterations we should have reached the max value
+	for i := uint64(1); i <= c.HappyPathRounds+uint64(tc.maxExponent)*2; i += 1 {
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply consensus/hotstuff/pacemaker/timeout/controller_test.go:115]: nice

File: consensus/hotstuff/integration/liveness_test.go [hunk 7]
```diff
@@ -217,3 +218,69 @@ func Test1TimeoutOutof5Instances(t *testing.T) {
...
+			err := in.Run()
+			require.True(t, errors.Is(err, errStopCondition))
+			wg.Done()
+		}(in)
+	}
+	wg.Wait()
+
+	// check that all instances have the same finalized block
+	ref := instances[0]
+	assert.Less(t, finalView-uint64(2*numPass+numFail), ref.forks.FinalizedBlock().View, "expect instance 0 should made enough progress, but didn't")
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply consensus/hotstuff/integration/liveness_test.go:296]: It's a legacy expression to make room for processed/finalized block view. I've updated those checks to be more precise using different stop condition.

File: consensus/hotstuff/integration/liveness_test.go [hunk 8]
```diff
@@ -217,3 +218,69 @@ func Test1TimeoutOutof5Instances(t *testing.T) {
...
+	finalizedViews := FinalizedViews(ref)
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply consensus/hotstuff/integration/liveness_test.go:298]: added this

File: consensus/hotstuff/integration/liveness_test.go [hunk 9]
```diff
@@ -211,8 +213,101 @@ func Test1TimeoutOutof5Instances(t *testing.T) {
...
+
+	// set up two instances which don't generate timeout objects
+	for n := healthyReplicas; n < healthyReplicas+replicasNotGeneratingTimeouts; n++ {
+		in := NewInstance(t,
+			WithRoot(root),
+			WithParticipants(participants),
+			WithLocalID(participants[n].NodeID),
+			WithTimeouts(timeouts),
+			WithStopCondition(ViewFinalized(finalView)),
+			WithOutgoingTimeoutObjects(BlockAllTimeoutObjects),
```
Comment (jordanschalm) [inline consensus/hotstuff/integration/liveness_test.go:277]: I'm still not convinced [this case](https://github.com/onflow/flow-go/pull/3201#discussion_r978120788) can't happen: 

> > We have 4 nodes in total where the TimeoutObjects from two of them are always discarded. Therefore, no TC can be constructed.
> 
> I'm not sure about this. 2/4 nodes will block outgoing TOs (`droppingReplicas`). However these nodes will still store their own TOs locally and may receive TOs from other healthy nodes. A `droppingReplica` could store its TO, then receive the TOs from both healthy nodes, and then be able to produce TC.
> 
> I think we also need to block inbound TOs on the two `droppingReplica`s.

Since we only block outbound TOs, a replica in `replicasNotGeneratingTimeouts` could:
* generate a TO locally, and store it
* receive 2 other TOs from the 2 nodes in `healthyReplicas`
* create a timeout certificate

I'm worried this test might be flaky, if the above condition can happen sometimes. If we block incoming TOs on the two `replicasNotGeneratingTimeouts`, then it's impossible.

**Review by jordanschalm — APPROVED** [review]
Nice work. I have one outstanding concern about a possible flaky test case (see [comment](https://github.com/onflow/flow-go/pull/3201/files#r979066355)), otherwise looks good!

File: consensus/hotstuff/integration/liveness_test.go [hunk 10]
```diff
@@ -217,3 +218,69 @@ func Test1TimeoutOutof5Instances(t *testing.T) {
 		assert.Equal(t, finalizedViews, FinalizedViews(instances[i]), "instance %d should have same finalized view as first instance")
 	}
 }
+
+// TestBlockDelayIsHigherThanTimeout tests an edge case where block delay is the same as round duration
+// which results in timing edge cases where some nodes generate timeout objects but don't yet form a TC but then receive
+// a proposal with newer QC.
+func TestBlockDelayIsHigherThanTimeout(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply consensus/hotstuff/integration/liveness_test.go]: I guess it could be a problem, I will modify it so byzantine replicas drop all timeouts, incoming and outgoing

[the remark this answers is not in the document]
Comment (durkmurder) [reply consensus/hotstuff/integration/liveness_test.go]: Strictly speaking for this test it's not an issue because healthy replicas will never construct a TC and eventually they will construct proposal with valid QC, so even if blocked nodes have moved forward after constructing TC, they still have already contributed into creation of QC by sending votes. For sake of clarity I will block incoming TO as well though.
